### PR TITLE
fix(MMAP/core/PathGenerator): Try to fix more water creatures

### LINKED ISSUE #13551: [Rather be Fishin'] Eels cannot swim upward to attack you
### Current Behaviour

Eels seem unable to swim upward and just follow you around at their current height.

The eels also moves very slowly. I assume this might be because the server calculation believes the eel is moving up at an angle and thus it's horizontal vector is lower.

https://user-images.githubusercontent.com/2512197/197927477-adb3d52c-11ab-42b5-8482-5d0d84d36e63.mp4

### Expected Blizzlike Behaviour

Eels should be able to swim upward.

### Source

_No response_

### Steps to reproduce the problem

1. Go to the northern lake in Terokkar Forest.
2. Attack an eel while swimming above the water.
3. The eel will not be able to come up and attack you.

### Extra Notes

Location: Outlands -> Terokkar Forest -> Silmyr Lake

The orb dropped by the eels also seems to be much higher than it should be (you are too far to loot ever when standing directly on the orb). See https://youtu.be/9nOWoaV-PYc?t=124 for expected behavior.

### AC rev. hash/commit

Azerothcore rev. 48fccc851d84+ 2022-04-06

### Operating system

Linux x64 (Docker)

### Custom changes or Modules

- Auction house

### PR DESCRIPTION
**Pull request by MikaMauger**
This reverts commit 66e6d331166b18c996cda6b58bca3131b524d4bc.

## Changes Proposed:
- Fix a lot of water creatures by reverting aforementioned PR, recast doesn't handle very well underwater terrain, above water terrain and water itself as connected, resulting in creatures using weird path or not be able to compute a path at all
- Check [TrinityCore PR](https://github.com/TrinityCore/TrinityCore/pull/28392) for more information
- Water creatures should follow you in a more straightforward way.
- Water creatures should continue to attack you even if you have waterwalk and/or are jumping in water (which was partially fixed by #12112, thanks to @UltraNix )

## Issues Addressed:

- Closes #13551
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/6784
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/12411
- Closes https://github.com/chromiecraft/chromiecraft/issues/3793

## Tests Performed:

- Before: path is wrong
WoWScrnShot_110522_112328
- After: path is ok
WoWScrnShot_110522_112601
- Before: can't even find a path
WoWScrnShot_110522_114143
- Before: When attacking from range, he gets stuck then evades/resets
WoWScrnShot_110522_113825
- After: path is ok
WoWScrnShot_110522_112713

If you want to see what happens in RecastDemo:
Capture_decran_2022-11-01_233355
Capture_decran_2022-11-01_233408

## How to Test the Changes:

1. `.go xyz -9243.936523 -2453.144531 56.821400` and do range attacks on murlocs from beach when they are in water
2. `.go xyz -9250.759766 -1182.086792 67.439095` and do range attacks on murlocs from beach when they are in water
3. A lot more locations where there is a river or a beach/coast

## Known Issues and TODO List:

- We need to find a way to handle underwater floor correctly, not very useful for 3.3.5, but definitely needed for cataclysm zones like Vashj'ir, which is not the target for AzerothCore anyway
- This PR doesn't fix the abrupt transitions between movement/attacks of water NPCs / hiccups.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by MikaMauger** [comment]
You'll need to regenerate mmaps in order to get the fix.

**Comment by pklloveyou** [comment]
image
It's really much better

**Comment by pklloveyou** [comment]
image
They will also follow the players from the "water surface" to the "water“

**Comment by acidmanifesto** [comment]
i'm going to mention this so there is no rush with people thinking this is some golden unicorn end all. 
There is a issue with this.
As current where it stands is water mobs that can walk on land can not be pulled from the water to land. 
They end up evading instead.
Maybe we will have less issues then current. or maybe we will have more issues reported over time.
Weigh the Pro's and Con's first and see if this can be adjusted to fix the issue, if not then a internal discussion will be needed in either correcting the pr, taking it as is, or closing it.

**Comment by acidmanifesto** [comment]
Also in reference to a screen shot earlier. NPC is swimming above the water surface.
image

**Comment by Gultask** [comment]
> Also in reference to a screen shot earlier. NPC is swimming above the water surface.

I think that's the normal running animation

**Comment by Gultask** [comment]
If anyone can. Please try:

1. Checking Gazh'ranka
`.additem 19975 5`
`.additem 19974`
`.go xyz -11695 -1770 13 309` Teleports to Pagle's Pointe
Use item, try to pull from the land in one instance, see the whole path taken in the other.

2. This issue: https://github.com/azerothcore/azerothcore-wotlk/issues/13551

**Comment by MikaMauger** [comment]
1. Gazh'ranka
- before:
https://user-images.githubusercontent.com/21247853/200922592-f1404770-2ea2-40b6-b593-6023cbd670cb.mp4
- after:
https://user-images.githubusercontent.com/21247853/200922636-8130e0d7-4907-47a3-baff-908110819e8b.mp4
2. Eels
- before:
https://user-images.githubusercontent.com/21247853/200923601-540d3a81-7a68-4fdc-8764-da378d124208.mp4
- after:
https://user-images.githubusercontent.com/21247853/200923640-96820deb-48d2-4d01-9cc3-09f3a26d996d.mp4
3. Water elem
- before:
https://user-images.githubusercontent.com/21247853/200923768-6b231bbf-ab32-4c47-aa53-0caa947b0435.mp4
- after:
https://user-images.githubusercontent.com/21247853/200923802-d8b56d9a-4a96-4f08-b870-d35c458302af.mp4
4. Snapjaw
- before:
https://user-images.githubusercontent.com/21247853/200923850-7bf4db03-1e84-46a7-a9c7-9299428b5846.mp4
- after:
https://user-images.githubusercontent.com/21247853/200923857-52099d9f-a71d-453d-9473-71d7173c4a89.mp4

**Comment by MikaMauger** [comment]
> > Also in reference to a screen shot earlier. NPC is swimming above the water surface.
> 
> I think that's the normal running animation

Indeed, he's just running here. The water effect on his feet always appear until NPC is fully on land.

**Comment by pklloveyou** [comment]
Feels better than before

**Review by UltraNix — APPROVED** [review]

**Comment by Gultask** [comment]
I've been testing for quite a while now and everything I can think of is much better than what it was. No regressions that I can see.
None of the problems reported in TC, either.

File: src/server/game/Movement/MovementGenerators/PathGenerator.cpp [hunk 1]
```diff
@@ -209,25 +209,30 @@ void PathGenerator::BuildPolyPath(G3D::Vector3 const& startPos, G3D::Vector3 con
...
-        bool isFarUnderWater = startFarFromPoly ? isUnderWaterStart : isUnderWaterEnd;
+        auto liquidDataStart = _source->GetMap()->GetLiquidData(_source->GetPhaseMask(), startPos.x, startPos.y, startPos.z, _source->GetCollisionHeight(), MAP_ALL_LIQUIDS);
+        auto liquidDataEnd = _source->GetMap()->GetLiquidData(_source->GetPhaseMask(), endPos.x, endPos.y, endPos.z, _source->GetCollisionHeight(), MAP_ALL_LIQUIDS);
 
+        bool startUnderWaterEndInWater = liquidDataStart.Status == LIQUID_MAP_UNDER_WATER &&
+                                         (liquidDataEnd.Status & MAP_LIQUID_STATUS_IN_CONTACT) != 0;
+        bool startInWaterEndUnderWater = (liquidDataStart.Status & MAP_LIQUID_STATUS_IN_CONTACT) != 0 &&
+                                         liquidDataEnd.Status == LIQUID_MAP_UNDER_WATER;
+        bool waterPath = startUnderWaterEndInWater || startInWaterEndUnderWater;
+
```
Comment (Nefertumm) [inline src/server/game/Movement/MovementGenerators/PathGenerator.cpp:222]:
```suggestion

```

**Comment by Nefertumm** [comment]
Removed blank space causing issues on CI

**Comment by smellbee** [comment]
when this will be merged?

**Comment by StraysFromPath** [comment]
I have tested this edit and it is a huge improvement fixing murlocs across darkshore and westfall.

**Comment by heyitsbench** [comment]
What is the status on this?

**Comment by Nyeriah** [comment]
> What is the status on this?

It's going to be merged after the holidays.

**Comment by Samsequel** [comment]
@MikaMauger 

I updated the MMAP version so the core will complain when old files are used

**Comment by Gultask** [comment]
Thank you **very** much for the amazing fix, @MikaMauger !

**Comment by Samsequel** [comment]
Yes its looking good so far :)

**Pull request merged**